src/sandbox/interpolation/scripIO.cc

```diff
@@ -0,0 +1,264 @@
+};
+
+
+SparseMatrixStorage ScripIO::read(const std::string& matrix_name) {
+#if ATLAS_HAVE_NETCDF == 0
+    ATLAS_THROW_EXCEPTION("Cannot read SCRIP files: Atlas not compiled with NetCDF support");
+#else
+    Size Nr  = 0;
+    Size Nc  = 0;
+    Size nnz = 0;
+
+    std::vector<Index> ia;
+    std::vector<Index> ja;
+    std::vector<Scalar> a;
+
+    // read SCRIP file
+    try {
+        netCDF::NcFile f(matrix_name, netCDF::NcFile::read);
+
+        Nr  = f.getDim("dst_grid_size").getSize();
+        Nc  = f.getDim("src_grid_size").getSize();
+        nnz = f.getDim("num_links").getSize();
+
+        ATLAS_ASSERT(Nr > 0);
+        ATLAS_ASSERT(Nc > 0);
+        ATLAS_ASSERT(nnz > 0);
+
+        auto var_ia = f.getVar("dst_address");
+        ATLAS_ASSERT(var_ia.getDimCount() == 1 && var_ia.getDim(0).getSize() == nnz);
+        ia.resize(nnz);  // NOTE: not compressed
+        var_ia.getVar(ia.data());
+
+        auto var_ja = f.getVar("src_address");
+        ATLAS_ASSERT(var_ja.getDimCount() == 1 && var_ja.getDim(0).getSize() == nnz);
+        ja.resize(nnz);
+        var_ja.getVar(ja.data());
+
+        auto var_a = f.getVar("remap_matrix");
+        ATLAS_ASSERT(var_a.getDimCount() == 2 && var_a.getDim(0).getSize() == nnz && var_a.getDim(1).getSize() == 1);
+        a.resize(nnz);
+        var_a.getVar(a.data());
+    }
+    catch (netCDF::exceptions::NcException& e) {
+        throw eckit::SeriousBug("SCRIP reading failed", Here());
+    }
+
+    // matrix conversion to 0-based indexing, sorting and compression
+    {
+        std::vector<size_t> sorted(nnz);
+        for (size_t n = 0; n < nnz; ++n) {
+            sorted[n] = n;
+        }
+
+        std::sort(sorted.begin(), sorted.end(), [&](auto i, auto j) {
+            return ia[i] != ia[j] ? ia[i] < ia[j] : ja[i] != ja[j] ? ja[i] < ja[j] : i < j;
+        });
+
+        decltype(ia) ia_unsorted(Nr + 1, 0);
+        decltype(ja) ja_unsorted(nnz);
+        ia.swap(ia_unsorted);
+        ja.swap(ja_unsorted);
+
+        const auto base = static_cast<Index>(1);
+        for (size_t n = 0; n < nnz; ++n) {
+            auto r = ia_unsorted[sorted[n]] - base;
+            auto c = ja_unsorted[sorted[n]] - base;
+            ATLAS_ASSERT(0 <= r && r < Nr);
+            ATLAS_ASSERT(0 <= c && c < Nc);
+
+            ia[r + 1]++;
+            ja[n] = c;
+        }
+
+        ia_unsorted.clear();
+        ja_unsorted.clear();
+
+        for (size_t r = 0; r < Nr; ++r) {
+            ia[r + 1] += ia[r];
+        }
+
+        ATLAS_ASSERT(ia.front() == 0 && ia.back() == nnz);
+
+        decltype(a) a_unsorted(nnz);
+        a.swap(a_unsorted);
+
+        for (size_t n = 0; n < nnz; ++n) {
+            a[n] = a_unsorted[sorted[n]];
+            ATLAS_ASSERT(0. <= a[n] && a[n] <= 1.);
+        }
+    }
+
+    eckit::linalg::SparseMatrix eckit_matrix(new InPlaceAllocator(Nr, Nc, nnz, ia.data(), ja.data(), a.data()));
+    return atlas::linalg::make_sparse_matrix_storage(eckit_matrix);
```

Comment:
We don't need to create an eckit sparse matrix here:
```suggestion
    return atlas::linalg::SparseMatrixStorage{
        atlas::linalg::SparseMatrixView<Scalar,Index>{Nr, Nc, nnz, a.data(), ja.data(), ia.data())}};
```
Also this means we don't need the InPlaceAllocator class above.

Reply:
Nice point! I am avoiding now the 'InPlaceAllocator'.